packages/app-metadata-controller/src/AppMetadataController.ts

```diff
@@ -0,0 +1,185 @@
+import { BaseController } from '@metamask/base-controller';
+import type {
+  StateMetadata,
+  ControllerGetStateAction,
+  ControllerStateChangeEvent,
+  RestrictedMessenger,
+} from '@metamask/base-controller';
+
+// Unique name for the controller
+const controllerName = 'AppMetadataController';
+
+/**
+ * The options that AppMetadataController takes.
+ */
+export type AppMetadataControllerOptions = {
+  state?: Partial<AppMetadataControllerState>;
+  messenger: AppMetadataControllerMessenger;
+  currentMigrationVersion?: number;
+  currentAppVersion?: string;
+};
+
+/**
+ * The state of the AppMetadataController
+ */
+export type AppMetadataControllerState = {
+  currentAppVersion: string;
+  previousAppVersion: string;
+  previousMigrationVersion: number;
+  currentMigrationVersion: number;
+};
+
+/**
+ * Function to get default state of the {@link AppMetadataController}.
+ */
```

Comment:
A bit silly we have to provide a `@returns`, I know, but perhaps we can copy from the SampleGasPricesController:

``` suggestion
/**
 * Constructs the default {@link AppMetadataController} state. This allows
 * consumers to provide a partial state object when initializing the controller
 * and also helps in constructing complete state objects for this controller in
 * tests.
 *
 * @returns The default {@link AppMetadataController} state.
 */
```